### DISCUSSION TITLE
Morph Markers Inserted Inside <script> Tags

### DISCUSSION THREAD

**Discussion by kellerjmrtn** [livewire/livewire#8544] [7. Bugs (archived)]
### Livewire version

v3.5.0

### Laravel version

v10.48.12

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

1. Create a Livewire component with some data to loop through
```
class TestMorphMarkers extends Component
{
    public function render()
    {
        return view('livewire.test-morph-markers', [
            'data' => [
                'test' => 'some test data',
                'test2' => 'more test data',
            ],
        ]);
    }
}
```
2. Create a view which attempts to loop through this data and inject it into a javascript variable
```
<div>
    <script>
        const myJsArray = [
            @foreach ($data as $key => $value)
                {
                    label: $key,
                    value: $value,
                },
            @endforeach
        ];

        console.log(myJsArray);
    </script>
</div>
```
3. The resulting HTML appears as
```
<script>
    const myJsArray = [
        <!--[if BLOCK]><![endif]-->                {
                label: $key,
                value: $value,
            },
                        {
                label: $key,
                value: $value,
            },
        <!--[if ENDBLOCK]><![endif]-->
    ];

    console.log(myJsArray);
</script>
```
4. An error is thrown in console due to the injected morph markers
```
Uncaught SyntaxError: missing ] after element list
```

### Repository or Wirebox URL

https://livewire.laravel.com/docs/morphing

#### Similar Issues

- https://github.com/livewire/livewire/discussions/7196

#### Workaround

The situation in my code is of course more complex than the example in the above steps to reproduce, but it's possible to create the intended data structure in a native PHP array and store it in a variable, then use the `@js` blade directive to insert it. Doing so with the above example would produce
```
<div>
    <script>
        const myJsArray = @js($data);

        console.log(myJsArray);
    </script>
</div>
```
Which would log to conosle:
```
Object { test: "some test data", test2: "more test data" }
```

#### Other ideas

Rather than the nuclear option of setting
```
'inject_morph_markers' => false,
```
It may be helpful to be able to enable/disable it on a more granular level like say, on a per-component basis (`protected bool $injectMorphMarkers;` or the like) or with a blade directive (`@withoutMorphs`). I'm not sure how difficult these would be to implement

It may also be helpful to intentionally avoid outputting morph markers any time a `<script>` tag is being rendered? Since in general it's pretty much bound to break things, given it's inserting HTML comments into JS. Given they're injected using a regex, however, this may be difficult to do

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by lrljoe** [comment]
Maybe because your array isn't valid?  Where's the separation between the array items...

The code that you shared:
```
    <script>
        const myJsArray = [
            @foreach ($data as $key => $value)
                {
                    label: $key,
                    value: $value,
                }
            @endforeach
        ];

        console.log(myJsArray);
    </script>
```
needs a comma to separate the array items.

Why on **earth** are you trying to do it like that though?  

You've got $wire to pass values between the frontend and backend, or you can just json encode the array.

This is not a bug, when using **correctly formatted code**, it correctly injects morph markers in appropriate places.

**Reply by kellerjmrtn** [reply]
@lrljoe the missing comma was an issue with my example, I've gone back and added it in. The
```
Uncaught SyntaxError: missing ] after element list
```
Error still occurs in this example due to the HTML comment

As for why it's being done this way, I'm using Livewire to control some form elements through `wire.model` which set some params to pull data from the DB. When a param is updated, the resulting data is then sent to a third party JS plugin embedded into the component that generates charts/graphs using the data. This may not be the ideal way to do things, but it worked in livewire v2 and I'm attempting to upgrade to v3

I don't *think* using `$wire` is practical for this situation, although I could be wrong. I do think using json encode or `@js` is the simplest solution moving forward, so I'll be rewriting my current Livewire code to work like that instead

As for using "correctly formatted code", are you essentially saying that we should **never** intermingle blade directives (`@if`, `@foreach`, etc) inside `<script>` tags? Or as in the linked discussion #7196, inside `<textarea>` tags? That there's just some places you should never be using these directives, at least in the context of Livewire? While it seems like in my case and the linked case, alternative methods/workarounds were found, I do think it may still be helpful to keep Livewire from inserting these HTML comments into contexts where they're invalid or create syntax errors, without having to completely disable the feature altogether using the `inject_morph_markers` config setting

**Reply by kellerjmrtn** [reply]
Perhaps a simple addition to the docs somewhere under https://livewire.laravel.com/docs/morphing#injecting-morph-markers could be a good idea? Something like (paraphrasing) "It's advised not to use conditionals inside script tags, as Livewire will generate morph markers around them, resulting in syntax errors"

I did notice the addition of `wire:replace` a few weeks ago in #8403. It contains a mention in the docs https://github.com/livewire/livewire/blob/main/docs/morph.md#bypassing-morphing which hasn't yet been released on the public docs https://livewire.laravel.com/docs/morphing. However this also did not seem to fix my issue, as it doesn't stop the morph markers from being generated, just ignores morphing altogether (by replacing the entire HTML instead of just the diff)

**Comment by goldmerc** [comment]
I just ran into this issue as well whilst updating a project from Livewire 2 to 3. In my case it was a simple @if, so I could resolve it by using curly braces instead. 

It does seem that the docs should mention that using blade directives inside script tags will fail due to the injected morph markers.

**Comment by CodyPChristian** [comment]
Ran into this too as mentioned in this related post https://github.com/livewire/livewire/discussions/7389#discussioncomment-10249691
